File: ios/Runner/YuvJpegConverter.swift
import Flutter
import UIKit
import CoreImage
import CoreGraphics

class YuvJpegConverter: NSObject, FlutterPlugin {
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "yuv_jpeg_converter", binaryMessenger: registrar.messenger())
        let instance = YuvJpegConverter()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "convert":
            guard let args = call.arguments as? [String: Any],
                  let imageData = args["image"] as? [String: Any],
                  let quality = args["quality"] as? Int else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid arguments", details: nil))
                return
            }
            
            do {
                let jpegData = try convertYuvToJpeg(imageData: imageData, quality: quality)
                result(jpegData)
            } catch {
                result(FlutterError(code: "CONVERSION_ERROR", message: "Failed to convert YUV to JPEG: \(error)", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func convertYuvToJpeg(imageData: [String: Any], quality: CGFloat) throws -> Data {
        guard let width = imageData["width"] as? Int,
              let height = imageData["height"] as? Int,
              let yPlane = imageData["yPlane"] as? FlutterStandardTypedData,
              let uPlane = imageData["uPlane"] as? FlutterStandardTypedData,
              let vPlane = imageData["vPlane"] as? FlutterStandardTypedData else {
            throw ConversionError.invalidData
        }
        
        let yRowStride = imageData["yRowStride"] as? Int ?? width
        let uRowStride = imageData["uRowStride"] as? Int ?? width / 2
        let vRowStride = imageData["vRowStride"] as? Int ?? width / 2
        let uPixelStride = imageData["uPixelStride"] as? Int ?? 1
        let vPixelStride = imageData["vPixelStride"] as? Int ?? 1
        
        // Convert YUV420 to RGB
        let rgbData = yuv420ToRgb(
            yPlane: yPlane.data,
            uPlane: uPlane.data,
            vPlane: vPlane.data,
            width: width,
            height: height,
            yRowStride: yRowStride,
            uRowStride: uRowStride,
            vRowStride: vRowStride,
            uPixelStride: uPixelStride,
            vPixelStride: vPixelStride
        )
        
        // Create CGImage from RGB data
        guard let colorSpace = CGColorSpace(name: CGColorSpace.sRGB),
              let context = CGContext(
                data: rgbData,
                width: width,
                height: height,
                bitsPerComponent: 8,
                bytesPerRow: width * 3,
                space: colorSpace,
                bitmapInfo: CGImageAlphaInfo.none.rawValue
              ),
              let cgImage = context.makeImage() else {
            throw ConversionError.imageCreationFailed
        }
        
        // Convert to JPEG
        guard let jpegData = cgImage.jpegData(compressionQuality: quality) else {
            throw ConversionError.jpegCompressionFailed
        }
        
        return jpegData
    }
    
    private func yuv420ToRgb(
        yPlane: Data,
        uPlane: Data,
        vPlane: Data,
        width: Int,
        height: Int,
        yRowStride: Int,
        uRowStride: Int,
        vRowStride: Int,
        uPixelStride: Int,
        vPixelStride: Int
    ) -> UnsafeMutablePointer<UInt8> {
        let rgbData = UnsafeMutablePointer<UInt8>.allocate(capacity: width * height * 3)
        
        for y in 0..<height {
            for x in 0..<width {
                let yIndex = y * yRowStride + x
                let uvY = y / 2
                let uvX = x / 2
                let uIndex = uvY * uRowStride + uvX * uPixelStride
                let vIndex = uvY * vRowStride + uvX * vPixelStride
                
                let yVal = Int(yPlane[yIndex])
                let uVal = Int(uPlane[uIndex]) - 128
                let vVal = Int(vPlane[vIndex]) - 128
                
                // YUV to RGB conversion
                var r = yVal + (1.370705 * Double(vVal)).rounded()
                var g = yVal - (0.698001 * Double(vVal)).rounded() - (0.337633 * Double(uVal)).rounded()
                var b = yVal + (1.732446 * Double(uVal)).rounded()
                
                r = max(0, min(255, r))
                g = max(0, min(255, g))
                b = max(0, min(255, b))
                
                let rgbIndex = (y * width + x) * 3
                rgbData[rgbIndex] = UInt8(r)
                rgbData[rgbIndex + 1] = UInt8(g)
                rgbData[rgbIndex + 2] = UInt8(b)
            }
        }
        
        return rgbData
    }
}

enum ConversionError: Error {
    case invalidData
    case imageCreationFailed
    case jpegCompressionFailed
}

extension CGImage {
    func jpegData(compressionQuality: CGFloat) -> Data? {
        guard let data = NSMutableData() as CFMutableData?,
              let destination = CGImageDestinationCreateWithData(data, kUTTypeJPEG, 1, nil) else {
            return nil
        }
        
        CGImageDestinationAddImage(destination, self, [kCGImageDestinationLossyCompressionQuality: compressionQuality] as CFDictionary)
        
        guard CGImageDestinationFinalize(destination) else {
            return nil
        }
        
        return data as Data
    }
}

File: android/app/src/main/java/com/example/jamal_test/YuvJpegConverter.kt
package com.example.jamal_test

import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.media.Image
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

class YuvJpegConverter: FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "yuv_jpeg_converter")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "convert" -> {
                try {
                    val imageData = call.argument<Map<String, Any>>("image")
                    val quality = call.argument<Int>("quality") ?: 75
                    
                    if (imageData == null) {
                        result.error("INVALID_ARGUMENT", "Image data is required", null)
                        return
                    }
                    
                    val jpegBytes = convertYuvToJpeg(imageData, quality)
                    result.success(jpegBytes)
                } catch (e: Exception) {
                    result.error("CONVERSION_ERROR", "Failed to convert YUV to JPEG: ${e.message}", null)
                }
            }
            else -> result.notImplemented()
        }
    }

    private fun convertYuvToJpeg(imageData: Map<String, Any>, quality: Int): ByteArray {
        val width = imageData["width"] as Int
        val height = imageData["height"] as Int
        
        val yPlane = imageData["yPlane"] as ByteArray
        val uPlane = imageData["uPlane"] as ByteArray
        val vPlane = imageData["vPlane"] as ByteArray
        
        val yRowStride = imageData["yRowStride"] as Int
        val uRowStride = imageData["uRowStride"] as Int
        val vRowStride = imageData["vRowStride"] as Int
        
        val uPixelStride = imageData["uPixelStride"] as Int
        val vPixelStride = imageData["vPixelStride"] as Int
        
        // Convert YUV420 to NV21 format (required by YuvImage)
        val nv21 = yuv420ToNv21(
            yPlane, uPlane, vPlane,
            width, height,
            yRowStride, uRowStride, vRowStride,
            uPixelStride, vPixelStride
        )
        
        // Create YuvImage and compress to JPEG
        val yuvImage = YuvImage(nv21, ImageFormat.NV21, width, height, null)
        val outputStream = ByteArrayOutputStream()
        
        yuvImage.compressToJpeg(Rect(0, 0, width, height), quality, outputStream)
        
        return outputStream.toByteArray()
    }

    private fun yuv420ToNv21(
        yPlane: ByteArray,
        uPlane: ByteArray,
        vPlane: ByteArray,
        width: Int,
        height: Int,
        yRowStride: Int,
        uRowStride: Int,
        vRowStride: Int,
        uPixelStride: Int,
        vPixelStride: Int
    ): ByteArray {
        val nv21 = ByteArray(width * height * 3 / 2)
        
        // Copy Y plane
        var yIndex = 0
        var nv21Index = 0
        for (row in 0 until height) {
            System.arraycopy(yPlane, yIndex, nv21, nv21Index, width)
            yIndex += yRowStride
            nv21Index += width
        }
        
        // Interleave U and V planes
        var uIndex = 0
        var vIndex = 0
        for (row in 0 until height / 2) {
            for (col in 0 until width / 2) {
                nv21[nv21Index++] = vPlane[vIndex]
                nv21[nv21Index++] = uPlane[uIndex]
                uIndex += uPixelStride
                vIndex += vPixelStride
            }
            uIndex += uRowStride - (width / 2) * uPixelStride
            vIndex += vRowStride - (width / 2) * vPixelStride
        }
        
        return nv21
    }
}
